cirq/contrib/acquaintance/executor.py

```diff
@@ -0,0 +1,140 @@
+        is_acquaintance_strategy)
+from cirq.contrib.acquaintance.permutation import (
+        PermutationGate,
+        LogicalIndex,
+        LogicalGates, LogicalMapping)
+from cirq.contrib.acquaintance.strategy import (
+        expose_acquaintance_gates)
+
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    from typing import Callable, List, DefaultDict
+
+class ExecutionStrategy(metaclass=abc.ABCMeta):
+    """An execution strategy tells StrategyExecutor how to execute an
+    acquaintance strategy, i.e. what gates to implement at the available
+    acquaintance opportunities."""
+
+    keep_acquaintance = False
+
+    @abc.abstractproperty
+    def initial_mapping(self) -> LogicalMapping:
+        pass
+
+    @abc.abstractmethod
+    def get_operations(self,
+                       indices: Sequence[LogicalIndex],
+                       qubits: Sequence[ops.QubitId]
+                       ) -> ops.OP_TREE:
+        """Gets the logical operations to apply to qubits."""
+        pass
+
+
+class StrategyExecutor(circuits.PointOptimizer):
+    """Executes an acquaintance strategy."""
+
+    def __init__(self, execution_strategy: ExecutionStrategy) -> None:
+        self.execution_strategy = execution_strategy
+        self.mapping = execution_strategy.initial_mapping.copy()
+
+    def __call__(self, strategy: circuits.Circuit):
+        if not is_acquaintance_strategy(strategy):
+            raise TypeError('not is_acquaintance_strategy(strategy)')
+        expose_acquaintance_gates(strategy)
+        super().optimize_circuit(strategy)
+        return self.mapping.copy()
+
+    def optimization_at(self,
+                        circuit: circuits.Circuit,
+                        index: int,
+                        op: ops.Operation):
+        if (isinstance(op, ops.GateOperation) and
+                isinstance(op.gate, AcquaintanceOpportunityGate)):
+            logical_indices = tuple(self.mapping[q] for q in op.qubits)
+            logical_operations = self.execution_strategy.get_operations(
+                    logical_indices, op.qubits)
+            clear_span = int(not self.execution_strategy.keep_acquaintance)
+
+            return circuits.PointOptimizationSummary(
+                    clear_span=clear_span,
+                    clear_qubits=op.qubits,
+                    new_operations=logical_operations)
+
+        if (isinstance(op, ops.GateOperation) and
+                isinstance(op.gate, PermutationGate)):
+            op.gate.update_mapping(self.mapping, op.qubits)
+            return
+
+        raise TypeError('Can only execute a strategy consisting of gates that '
+                         'are instances of AcquaintanceOpportunityGate or '
+                         'PermutationGate.')
+
+class GreedyExecutionStrategy(ExecutionStrategy):
+    """A greedy execution strategy. When an acquaintance opportunity is
+    reached, all gates acting on those qubits in any order are inserted.
+
+        Args:
```

Comment:
Classes don't have arguments, they have attributes.

Is this supposed to be on init?

Reply:
Yes; fixed.